include/construction/ctx_generic.hpp

```diff
@@ -0,0 +1,240 @@
+          template <bool> typename rho_type,
+          bool require_initialization,
+          template <bool> typename bv_type>
+class ctx_generic {
+public:
+  ctx_generic() = default;
+
+  template<typename BvSizeArg>
+  ctx_generic(BvSizeArg&& size_arg,
+              uint64_t const levels,
+              rho_type<is_tree>&& rho,
+              uint64_t const shards = 1)
+      : hist_(levels, shards),
+        borders_(levels, shards),
+        zeros_(levels, 0),
+        bv_(levels, size_arg),
+        rho_(std::move(rho)),
+        levels_(levels) {}
+
+  uint64_t hist_size(uint64_t const level) {
+    return 1ull << level;
+  }
+
+  auto hist_at_shard(uint64_t const shard) {
+    DCHECK(hist_array::is_not_leveled);
+    return hist_.get(shard, 0 /* TODO: 0 is arbitrary here*/);
+  }
+  auto hist_at_shard(uint64_t const shard) const {
+    DCHECK(hist_array::is_not_leveled);
+    return hist_.get(shard, 0 /* TODO: 0 is arbitrary here*/);
+  }
+
+  auto hist_at_level(uint64_t const level) {
+    DCHECK(hist_array::is_not_sharded);
+    return hist_.get(0 /* TODO: 0 is arbitrary here*/, level);
+  }
+  auto hist_at_level(uint64_t const level) const {
+    DCHECK(hist_array::is_not_sharded);
+    return hist_.get(0 /* TODO: 0 is arbitrary here*/, level);
+  }
+
+  auto borders_at_shard(uint64_t const shard) {
+    DCHECK(borders_array::is_not_leveled);
+    return borders_.get(shard, 0 /* TODO: 0 is arbitrary here*/);
+  }
+  auto borders_at_shard(uint64_t const shard) const {
+    DCHECK(borders_array::is_not_leveled);
+    return borders_.get(shard, 0 /* TODO: 0 is arbitrary here*/);
+  }
+
+  auto borders_at_level(uint64_t const level) {
+    DCHECK(borders_array::is_not_sharded);
+    return borders_.get(0 /* TODO: 0 is arbitrary here*/, level);
+  }
+  auto borders_at_level(uint64_t const level) const {
+    DCHECK(borders_array::is_not_sharded);
+    return borders_.get(0 /* TODO: 0 is arbitrary here*/, level);
+  }
+
+  auto hist_at(uint64_t const shard, uint64_t const level) {
+    return hist_.get(shard, level);
+  }
+  auto hist_at(uint64_t const shard, uint64_t const level) const {
+    return hist_.get(shard, level);
+  }
+  auto borders_at(uint64_t const shard, uint64_t const level) {
+    return borders_.get(shard, level);
+  }
+  auto borders_at(uint64_t const shard, uint64_t const level) const {
+    return borders_.get(shard, level);
+  }
+
+  uint64_t rho(size_t level, size_t i) {
+    return rho_.get(level, i);
+  }
+
+  void set_rho(size_t level, size_t i, uint64_t val) {
+    rho_.set(level, i, val);
+  }
+
+  static bool constexpr compute_zeros = !is_tree;
+  static bool constexpr compute_rho = rho_type<is_tree>::compute_rho;
+
+  // TODO: return span
+  std::vector<uint64_t>& zeros() {
+    return zeros_;
+  }
+
+  bv_type<require_initialization>& bv() {
+    return bv_;
+  }
+
+  bv_type<require_initialization> const& bv() const {
+    return bv_;
+  }
+
+  void discard_non_merge_data() {
```

Comment:
Unused functions can be removed, or is there a reason to keep it?

Reply:
Ah, that is still a TODO section actually - The `dd` algorithms call this to clean up data structures, but I want to move this method out of here anyway (and make it actually do something again).

Reply:
I added a TODO comment to it.

Reply:
I've replaced the empty method with a working implementation.